### ISSUE TITLE
IE11/Win7 ciphers ignored despite configured with `CipherSuitesPolicy` - please advice

### ISSUE THREAD

**Issue by nros** [dotnet/runtime#36456]
IE11/Win7 ciphers ignored despite configured with `CipherSuitesPolicy` - please advice

**Please advise**.

## Issue

IE11/Win7 can not connect to Kestrel server (from .NET Core 3.1) with TLS enabled.

IE11/Win7 supported ciphers are ignored, despite they are configured with `CipherSuitesPolicy` on Linux (CentOS 7 with openssl 1.1.1d). Only a subset of configured ciphers is used. 

## Why
We need to support IE11 on Win7, as this is a requirement for Office add-ins - at least we were told so by the Microsoft team checking our add-in. Office 2013 on Win7 utilized IE11 for running add-ins.

## Relevant code

We have already configured Kestrel to support TLS 1.2 and TLS 1.3 and allow ciphers required for IE11/Win7. (*TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256* and *TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384*).

The code is:
```C#
CipherSuitesPolicy CipherSuitesPolicy = new CipherSuitesPolicy
(
    new TlsCipherSuite[]
    {
        // Cipher suits as recommended by: https://wiki.mozilla.org/Security/Server_Side_TLS
        // Listed in preferred order.

        // From: https://en.internet.nl
        // High
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
        // Medium
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA256,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
        TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_AES_128_GCM_SHA256,
        TlsCipherSuite.TLS_AES_256_GCM_SHA384,
        TlsCipherSuite.TLS_CHACHA20_POLY1305_SHA256
    }
);


webHostBuilder.UseKestrel((builderContext, options) =>
{
    options.ConfigureHttpsDefaults(httpsOptions =>
    {
        httpsOptions.SslProtocols = System.Security.Authentication.SslProtocols.Tls12 | System.Security.Authentication.SslProtocols.Tls13;

        httpsOptions.OnAuthenticate = (conContext, sslAuthOptions) =>
        {
            sslAuthOptions.CipherSuitesPolicy = CipherSuitesPolicy;
        };
    }

    X509Certificate2 certificate = null; /* loaded from filesystem in real code - a bit lengthy */
    int webPort = _configuration.GetValue<int>("WebServer:Webort");
    options.Listen(IPAddress.Any, webPort, listenOptions => listenOptions.UseHttps(certificate));
}

webHostBuilder.UseIISIntegration();
/* ... */
```

## Performed tests

Nonetheless, tests reveal that only a limited number of supported ciphers are supported by the server:
* see [SSLLabs test page](https://www.ssllabs.com/ssltest/analyze.html?d=demo.smint.io&hideResults=on)
* tested with [sslyze](https://github.com/nabla-c0d3/sslyze) `sslyze  --tlsv1_2  demo.smint.io:443`
    ```
     SCAN RESULTS FOR DEMO.SMINT.IO:443 - 13.69.28.15
     ------------------------------------------------

     * TLS 1.2 Cipher suites:
         Attempted to connect using 158 cipher suites.

         The server accepted the following 3 cipher suites:
            TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256       256       ECDH: x25519 (253 bits)
            TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384             256       ECDH: prime256v1 (256 bits)
            TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             128       ECDH: prime256v1 (256 bits)

         The group of cipher suites supported by the server has the following properties:
           Forward Secrecy                    OK - Supported
           Legacy RC4 Algorithm               OK - Not Supported

         The server has no preferred cipher suite.
    ```
* WireShark revealed, that Kestrel server just closes/drops TLS connection without any *Server Hello* answer to the *Client Hello* received from IE 11/Win7 client. Edge and Chrome receive proper *Server Hello* answer with a selected cipher.
* openssl on the Linux machine supports these ciphers
    ```
    > openssl ciphers -V
    ...
    0xC0,0x2C - ECDHE-ECDSA-AES256-GCM-SHA384 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(256) Mac=AEAD
    0xC0,0x2B - ECDHE-ECDSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(128) Mac=AEAD
    ...
    ```

**Comment by wfurt** [comment]
I think @vcsjones bump to something similar recently and may have some suggestions. From the description I assume you running Kestrel server on Linux? Either directly or in container? 

It would be helpful if you can post Wireshark capture or find way how to mimic it with OpenSSL.

**Comment by nros** [comment]
@wfurt Kestrel is running directly on Linux server - no container. 

Here are some WireShark captures:
* IE11 on Windows 7 - https://www.smint.io/downloads/github-issues/dotnet/runtime/36456/capture-demo.smint.io-ie11-win7.pcapng
* Chromium with TLS 1.3 on MacOs - https://www.smint.io/downloads/github-issues/dotnet/runtime/36456/capture-demo.smint.io-chrome.pcapng
* Firefox limited to TLS 1.2 only on MacOS - https://www.smint.io/downloads/github-issues/dotnet/runtime/36456/capture-demo.smint.io-firefox-tls12.pcapng

**Comment by nros** [comment]
More scrutiny revealed:

* IE11/Win7 is creating an initial "Client Hello" request with TLS 1.2, but then aborts, restarts and uses TLS 1.0
* Since TLS 1.0 is deactivated for good reasons, server closes connection because of unsuported protocol.

To me it is yet unknown, why IE11 aborts and restarts the session. According to this post on StackOverflow, it might happen in case the server certificate used MD5 - which is not the case. Server certificate uses SHA256.

https://serverfault.com/questions/741706/bug-with-tls-1-2-in-internet-explorer-11-windows-7

**Comment by nros** [comment]
Problem is with openssl [`ssl_set_mask()`](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1c/ssl/ssl_lib.c#L3234). The algorithms supported on both sides are masked because somehow `SSL_aECDSA` is not applied to the mask of allowed types. This mask is used in `ssl3_choose_cipher()` in file "s3_lib.c" to choose a common cipher. Unfortunately debug shows, that the algorithms are ruled out (ok == 0):
```
ok:[alg_k:alg_a:mask_k:mask_a]c:cipher-name
0:[00000004:00000008:000000CD:00000015]0x7ff3303a5d70:ECDHE-ECDSA-AES256-GCM-SHA384
0:[00000004:00000008:000000CD:00000015]0x7ff3303a5d20:ECDHE-ECDSA-AES128-GCM-SHA256
```

The mask_a is composed of: `SSL_kRSA`, `SSL_kRSA`, `SSL_aPSK`, but needs `SSL_aECDSA`, too. 
Can anybody help to shed light on this?

**Comment by wfurt** [comment]
Look at openssl.conf and #30767
also @bartonjs  may have some insight.

**Comment by bartonjs** [comment]
The server certificate has an RSA public key, so the ECDSA ciphersuites are ineligible.

At least, that's why the external tests show TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, but not TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384.

**Comment by nros** [comment]
@wfurt thanks, I have already tried to overcome the issue with settings these values in `openssl.cnf`. It did not work, though. I have verified that the file has been loaded by tweaking the openssl source and adding some STDERR output stating the path to the config file to be loaded.

@bartonjs Thanks for your hint. So the server certificate is the issue? I will try to research and find documentation about public key requirements to support IE11 Win7.

So, if the server key is the issue that blocks IE11 in Win7, than this issue can be closed, as this is not an issue with Kestrel. I will do after a have found some more docs and will post them for everybody stumbling on the same issue.

**Comment by nros** [comment]
As was totally unaware that the possible algorithms depend on the server certificate. I am still confused about that. 

By the way, the other two by IE11 supported strong ciphers are: *DHE-RSA-AES256-GCM-SHA384* and  *DHE-RSA-AES128-GCM-SHA256*. But these are rejected too.
```
ok:[alg_k:alg_a:mask_k:mask_a]c:cipher-name
0:[00000002:00000001:000000CD:00000015]0x7ff3303a4d80:DHE-RSA-AES256-GCM-SHA384
0:[00000002:00000001:000000CD:00000015]0x7ff3303a4d30:DHE-RSA-AES128-GCM-SHA256
```
There the mask_k lacks `SSL_kDHE`. So, openssl library does not use/create DHE temporary key with RSA signature. Does this depend on certificate too?

To summarize, DHE or ECDHE based keys can not be used if the server certificate uses RSA? So, choose your CA wisely. 

see:
* https://security.stackexchange.com/questions/45339/tls-1-2-server-certificate-and-signature-algorithms



My solution was, to enable the weak key `TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384`

**Comment by vcsjones** [comment]
In TLS 1.2 and below the RSA / ECDSA part of the suite name is the signature algorithm of the certificate.

[SChannel in Windows 7][1] oddly enough does not support RSA + ECDHE + AES-GCM. Windows 7 there only supports DHE and static key RSA with AES-GCM. 

> There the mask_k lacks SSL_kDHE

So, openssl library does not use/create DHE temporary key with RSA signature. Does this depend on certificate too?

Windows 7 supports DHE + AES-GCM but many server operators have turned off DHE-based suites. They are slow to generate ephmeral keys, and LogJam attack showed that many servers misconfigured their Diffie-Hellman parameters or accepted weak ones.

> To summarize, DHE or ECDHE based keys can not be used if the server certificate uses RSA?

This is a limitation of Windows 7, not TLS. ECDHE + RSA + AES-GCM should work fine on Windows 10.

[1]: https://docs.microsoft.com/en-us/windows/win32/secauthn/tls-cipher-suites-in-windows-7

**Comment by nros** [comment]
@vcsjones  thanks for clarification. Apparently, I have learned a lot from this issue :)

As promised to all future visitors, here is a link to [RFC 5246 page 49](https://tools.ietf.org/html/rfc5246#page-49), which contains more information about requirements for the certificates.

**Comment by bartonjs** [comment]
Seems like the questions have been answered, so closing the issue.
